numpy/core/src/umath/ufunc_object.c

```diff
@@ -3274,38 +2850,52 @@ PyUFunc_GenericFunction_int(PyUFuncObject *ufunc,
 fail:
     NPY_UF_DBG_PRINT1("Returning failure code %d\n", retval);
     for (i = 0; i < nop; ++i) {
-        Py_XDECREF(op[i]);
-        op[i] = NULL;
         Py_XDECREF(dtypes[i]);
         Py_XDECREF(arr_prep[i]);
     }
-    Py_XDECREF(type_tup);
-    Py_XDECREF(extobj);
-    Py_XDECREF(full_args.in);
-    Py_XDECREF(full_args.out);
-    Py_XDECREF(wheremask);
 
     return retval;
 }
 
 
-/*UFUNC_API*/
+/*UFUNC_API
+ * This generic function is called with the ufunc object, the arguments to it,
+ * and an array of (pointers to) PyArrayObjects which are NULL.
+ *
+ * 'op' is an array of at least NPY_MAXARGS PyArrayObject *.
+ */
 NPY_NO_EXPORT int
 PyUFunc_GenericFunction(PyUFuncObject *ufunc,
         PyObject *args, PyObject *kwds, PyArrayObject **op)
 {
     /* NumPy 1.19, 2020-01-24 */
     if (DEPRECATE(
-            "PyUFunc_GenericFunction() C-API function is deprecated "
-            "and expected to be removed rapidly. If you are using it (i.e. see "
-            "this warning/error), please notify the NumPy developers. "
-            "As of now it is expected that any use case is served better by "
-            "the direct use of `PyObject_Call(ufunc, args, kwargs)`. "
-            "PyUFunc_GenericFunction function has slightly different "
-            "untested behaviour.") < 0) {
+                "PyUFunc_GenericFunction() C-API function is deprecated "
+                "and expected to be removed rapidly. If you are using it (i.e. see "
+                "this warning/error), please notify the NumPy developers. "
+                "As of now it is expected that any use case is served better by "
+                "the direct use of `PyObject_Call(ufunc, args, kwargs)`. "
+                "PyUFunc_GenericFunction function has slightly different "
+                "untested behaviour.") < 0) {
+        return -1;
+    }
+    if (ufunc == NULL) {
+        PyErr_SetString(PyExc_ValueError, "function not supported");
         return -1;
     }
-    return PyUFunc_GenericFunction_int(ufunc, args, kwds, op);
+    if (op == NULL) {
+        PyErr_SetString(PyExc_ValueError,
+                        "PyUFunc_GenericFunction() op must not be NULL.");
+        return -1;
+    }
+
+    PyObject *res = ufunc_generic_call_mps(ufunc, args, kwds, op);
```

Comment:
Renamed it to `_with_operands`, but you are right and I will make a PR after this to just remove the whole thing. Since nobody complained yet, I assume we can go ahead with disabling `PyUFunc_GenericFunction` for good.